optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/ConstraintSubTree.java

```diff
@@ -90,12 +90,22 @@ public RuleAssembler getRuleAssembler() {
     }
 
     public RuleAssembler getRuleAssembler(int totalExpectedGroupByCount) {
-        RuleAssembler builder = isJoin ? leftSubTree.getRuleAssembler(totalExpectedGroupByCount)
-                .join(rightSubTree.getRuleAssembler(totalExpectedGroupByCount), nodeList.get(0))
-                : RuleAssembler.from(variableFactory, nodeList.get(0), totalExpectedGroupByCount);
+        RuleAssembler assembler = getRuleAssemblerForFirstNode(totalExpectedGroupByCount);
         for (int i = 1; i < nodeList.size(); i++) {
-            builder = builder.andThen(nodeList.get(i));
+            ConstraintGraphNode nextNode = nodeList.get(i);
+            assembler = assembler.andThen(nextNode);
+        }
+        return assembler;
+    }
+
+    private RuleAssembler getRuleAssemblerForFirstNode(int totalExpectedGroupByCount) {
+        ConstraintGraphNode firstNode = nodeList.get(0);
+        if (isJoin) {
+            RuleAssembler left = leftSubTree.getRuleAssembler(totalExpectedGroupByCount);
+            RuleAssembler right = rightSubTree.getRuleAssembler(totalExpectedGroupByCount);
+            return left.join(right, firstNode);
+        } else {
+            return RuleAssembler.from(variableFactory, firstNode, totalExpectedGroupByCount);
         }
-        return builder;
```

Comment:
This is purely aesthetic. Was so much easier to debug, now that the condition isn't a one-liner.